StringIndexOutOfBoundsException when converting 'DecimalType' to 'DateTimeType'

I get the following error since I updated to M5:

```
2019-11-20 16:32:15.390 [ERROR] [ome.core.thing.link.ThingLinkManager] - Exception occurred while informing handler: String index out of range: 10

java.lang.StringIndexOutOfBoundsException: String index out of range: 10

	at java.lang.String.substring(String.java:1963) ~[?:1.8.0_222]

	at org.eclipse.smarthome.core.library.types.DateTimeType.<init>(DateTimeType.java:121) ~[?:?]

	at org.eclipse.smarthome.core.library.types.DecimalType.as(DecimalType.java:173) ~[?:?]

	at org.eclipse.smarthome.core.internal.items.ItemStateConverterImpl.convertToAcceptedState(ItemStateConverterImpl.java:64) ~[?:?]

	at org.eclipse.smarthome.core.thing.internal.profiles.ProfileCallbackImpl.sendUpdate(ProfileCallbackImpl.java:134) ~[?:?]

	at org.eclipse.smarthome.core.thing.internal.profiles.SystemDefaultProfile.onStateUpdateFromHandler(SystemDefaultProfile.java:53) ~[?:?]

	at org.eclipse.smarthome.core.thing.internal.CommunicationManager.lambda$9(CommunicationManager.java:467) ~[?:?]

	at org.eclipse.smarthome.core.thing.internal.CommunicationManager.lambda$11(CommunicationManager.java:487) ~[?:?]

	at java.lang.Iterable.forEach(Iterable.java:75) ~[?:1.8.0_222]

	at org.eclipse.smarthome.core.thing.internal.CommunicationManager.handleCallFromHandler(CommunicationManager.java:483) ~[?:?]

	at org.eclipse.smarthome.core.thing.internal.CommunicationManager.stateUpdated(CommunicationManager.java:465) ~[?:?]

	at org.eclipse.smarthome.core.thing.internal.ThingManagerImpl$1.stateUpdated(ThingManagerImpl.java:168) ~[?:?]

	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateState(BaseThingHandler.java:257) ~[?:?]

	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateState(BaseThingHandler.java:276) ~[?:?]

	at org.openhab.binding.kodi.internal.handler.KodiHandler.updateRating(KodiHandler.java:805) ~[?:?]

	at org.openhab.binding.kodi.internal.protocol.KodiConnection.requestPlayerItemUpdate(KodiConnection.java:672) ~[?:?]

	at org.openhab.binding.kodi.internal.protocol.KodiConnection.requestPlayerUpdate(KodiConnection.java:518) ~[?:?]

	at org.openhab.binding.kodi.internal.protocol.KodiConnection.updatePlayerStatus(KodiConnection.java:509) ~[?:?]

	at org.openhab.binding.kodi.internal.handler.KodiHandler.handleCommand(KodiHandler.java:306) ~[?:?]

	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.channelLinked(BaseThingHandler.java:203) ~[?:?]

	at org.eclipse.smarthome.core.thing.link.ThingLinkManager.lambda$0(ThingLinkManager.java:267) ~[?:?]

	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_222]

	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_222]

	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_222]

	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:1.8.0_222]

	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_222]

	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_222]

	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_222]
```
This error happens on startup or at bundle refresh.

Also it doesn't update some chanels after this error.

--- a/bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/library/types/DateTimeType.java
+++ b/bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/library/types/DateTimeType.java
@@ -104,10 +104,11 @@ public class DateTimeType implements PrimitiveType, State, Command {
                     date = parse("1970-01-01T" + zonedValue);
                 } catch (DateTimeParseException timeOnlyException) {
                     try {
-                        Long epoch = Long.valueOf(zonedValue);
+                        long epoch = Double.valueOf(zonedValue).longValue();
+                        int length = (int) (Math.log10(epoch >= 0 ? epoch : epoch * -1) + 1);
                         Instant i;
                         // Assume that below 12 digits we're in seconds
-                        if (zonedValue.length() < 12) {
+                        if (length < 12) {
                             i = Instant.ofEpochSecond(epoch);
                         } else {
                             i = Instant.ofEpochMilli(epoch);

--- a/bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DecimalTypeTest.java
+++ b/bundles/org.openhab.core/src/test/java/org/eclipse/smarthome/core/library/types/DecimalTypeTest.java
@@ -121,6 +121,12 @@ public class DecimalTypeTest {
     public void testConversionToDateTimeType() {
         assertEquals(new DateTimeType("2014-03-30T10:58:47+0000"),
                 new DecimalType("1396177127").as(DateTimeType.class));
+        assertEquals(new DateTimeType("1969-12-31T23:59:59+0000"), new DecimalType("-1").as(DateTimeType.class));
+        assertEquals(new DateTimeType("1970-01-01T00:00:00+0000"), DecimalType.ZERO.as(DateTimeType.class));
+        assertEquals(new DateTimeType("1970-01-01T00:00:01+0000"), new DecimalType("1").as(DateTimeType.class));
+        assertEquals(new DateTimeType("1970-01-01T00:00:01+0000"), new DecimalType("1.0").as(DateTimeType.class));
+        assertEquals(new DateTimeType("1970-01-01T00:00:01+0000"), new DecimalType("1.5").as(DateTimeType.class));
+        assertEquals(new DateTimeType("1969-12-31T23:59:59+0000"), new DecimalType("-1.0").as(DateTimeType.class));
     }
 
     @Test

